solrmonitor/solrmonitor.go

```diff
@@ -150,8 +150,6 @@ func (c *SolrMonitor) GetCollectionState(name string) (*CollectionState, error)
 }
 
 func (c *SolrMonitor) doGetCollectionState(name string) (*CollectionState, error) {
-	c.mu.RLock()
```

Comment:
Are we sure this is really the problem? I took a look around and it seems that the only two callers for this acquire a read lock beforehand. Even though it is bad practice, the process won't deadlock if you are _only_ recursively read locking. For example, this code snippet returns perfectly fine:

```go
type locker struct {
	sync.RWMutex
}

func (l *locker) f(count int) {
	l.RLock()
	defer l.RUnlock()
	if count <= 0 {
		return
	}
	l.f(count - 1)
}

func TestWhatHappens(t *testing.T) {
	l := &locker{}
	l.f(10)
}
```

It's only a problem if your recursive flows are both read _and_ write locking, since you cannot acquire a write lock until all readers have relinquished and vice versa.

Reply:
if one thread is taking recursive lock and in-between other threads block on write lock. Then first thread won't be able to take read lock as after write-lock-block, golang doesn't allow read lock.

Reply:
ah that's correct, I misunderstood the behavior when both a reader and writer try to acquire. makes sense!